8144: Add rule to detect GC Inverted Parallelism

aptmac opened the pull request:
This PR addresses JMC-8144 [[0]](https://bugs.openjdk.org/browse/JMC-8144). There is a GC log analysis tool called Garbagecat [[1]](https://github.com/mgm3746/garbagecat/tree/main), which operates on GC logs and identifies performance issues and suggests potential solutions (similar to our automated analysis page). One metric Garbagecat flags is **"inverted parallelism"**, which takes a look at the sys+usr time of a parallel GC and compares it against the real (wall) time. If the parallelism is low, this can indicate that the parallel gc is operating at the efficiency of serial gc, and there could be room for improvement. This GC time information is now included in the jdk.GCCPUTime event added in JDK20+.

This PR requires support for the jdk.GCCPUTime event [[2]](https://bugs.openjdk.org/browse/JDK-8291753), so it also doubles as a PR towards JMC-7895 [[3]](https://bugs.openjdk.org/browse/JMC-7895).

[0] https://bugs.openjdk.org/browse/JMC-8144
[1] https://github.com/mgm3746/garbagecat/tree/main
[2] https://bugs.openjdk.org/browse/JDK-8291753
[3] https://bugs.openjdk.org/browse/JMC-7895

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed (1 review required, with at least 1 [Committer](https://openjdk.org/bylaws#committer))

### Issue
 * [JMC-8144](https://bugs.openjdk.org/browse/JMC-8144): Add rule to detect GC Inverted Parallelism (**Enhancement** - P4)


### Reviewers
 * [Marcus Hirt](https://openjdk.org/census#hirt) (@thegreystone - **Reviewer**)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jmc.git pull/528/head:pull/528` \
`$ git checkout pull/528`

Update a local copy of the PR: \
`$ git checkout pull/528` \
`$ git pull https://git.openjdk.org/jmc.git pull/528/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 528`

View PR using the GUI difftool: \
`$ git pr show -t 528`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jmc/pull/528.diff">https://git.openjdk.org/jmc/pull/528.diff</a>

</details>


### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jmc/pull/528#issuecomment-1785443516)

thegreystone commented:
Looks good. Perhaps a JDK 20 recording should be added to the tests so that some of the rule code is exercised?

aptmac commented:
> Looks good. Perhaps a JDK 20 recording should be added to the tests so that some of the rule code is exercised from a recording/events?

I've added two recordings here, taken on jdk21:
- **jdk21-inverted-parallelism.jfr**: has jdk.GCCPUTime enabled, and runs for ~10 seconds on a stress test application to force an instance of inverted parallelism
- **jdk21.jfr**: enables all events (minus the ones that have my system information), and runs on JMC itself for 1 minute

Also added a requirement on jdk.GCConfiguration for GcInvertedParallelismRule, and rebased ontop of master.

aptmac commented:
/integrate

franz1981 commented:
Nice @aptmac 👍👍
